### PR TITLE
Add support for new features on query results

### PR DESCRIPTION
**Pull request by RichardKeo**
Add support for a few features that will soon become available when retrieving query results:
  - specify the output columns
  - get a sample of the results
  - filter out some rows before retrieve the results
  - sort the results

### CODE REVIEW

File: dune_client/api/base.py [hunk 1]
```diff
@@ -83,6 +83,55 @@ def default_headers(self) -> Dict[str, str]:
...
+                col = '"' + column.replace('"', '\\"') + '"'
+                output.append(col)
+
+            params["columns"] = ",".join(output)
+        if sample_count is not None:
+            params["sample_count"] = sample_count
+        if filters is not None:
+            params["filters"] = filters
+        if sort_by is not None and len(sort_by) > 0:
+            params["sort_by"] = ",".join(sort_by)
```
Comment (TheEdgeOfRage) [inline dune_client/api/base.py:127]: Wouldn't you want to do the escaping here too?

File: dune_client/api/extensions.py [hunk 2]
```diff
@@ -47,41 +47,95 @@ def run_query(
...
         fetches and returns the results.
         Sleeps `ping_frequency` seconds between each status request.
         """
+        # Ensure we don't specify parameters that are incompatible:
+        assert (
+            # We are not sampling
+            sample_count is None
+            # We are sampling and don't use filters or pagination
+            or (batch_size is None and filters is None)
+        ), "sampling cannot be combined with filters or pagination"
```
Comment (TheEdgeOfRage) [inline dune_client/api/extensions.py:67]: Isn't this already covered by the assert on the build params function?

File: dune_client/api/extensions.py [hunk 3]
```diff
@@ -47,41 +47,95 @@ def run_query(
...
         fetches and the results in CSV format
         (use it load the data directly in pandas.from_csv() or similar frameworks)
         """
+        # Ensure we don't specify parameters that are incompatible:
+        assert (
+            # We are not sampling
+            sample_count is None
+            # We are sampling and don't use filters or pagination
+            or (batch_size is None and filters is None)
+        ), "sampling cannot be combined with filters or pagination"
```
Comment (TheEdgeOfRage) [inline dune_client/api/extensions.py:109]: Same

**Review by TheEdgeOfRage — COMMENTED** [review]
There's a few more places with asserts that I think are duplicate. Please check

File: dune_client/api/base.py (hunk 1 above)
Comment (TheEdgeOfRage) [reply dune_client/api/base.py:127]: Ah makes sense. We'll have to check for SQL injection then tho, right?

**Review by TheEdgeOfRage — APPROVED** [review]

File: dune_client/api/extensions.py [hunk 4]
```diff
@@ -47,41 +47,95 @@ def run_query(
...
+
+        if sample_count is not None:
+            limit = None
+        else:
+            limit = 10 or batch_size or MAX_NUM_ROWS_PER_BATCH
```
Comment (msf) [inline dune_client/api/extensions.py:72]: is `10 or ` there for local testing? looks like a bug?

File: dune_client/api/extensions.py [hunk 5]
```diff
@@ -114,6 +179,14 @@ def get_latest_result(
         :param max_age_hours: re-executes the query if result is older than max_age_hours
             https://dune.com/docs/api/api-reference/get-results/latest-results
```
Comment (msf) [inline dune_client/api/extensions.py:180]: update link to docs

File: dune_client/client_async.py [hunk 6]
```diff
@@ -372,15 +458,27 @@ async def refresh_into_dataframe(
...
+        filters: Optional[str] = None,
+        sort_by: Optional[List[str]] = None,
     ) -> ResultsResponse:
         """GET a page of results from Dune API for `job_id` (aka `execution_id`)"""
 
-        params = {
-            "limit": limit,
-            "offset": offset,
-        }
+        if sample_count is None and limit is None and offset is None:
```
Comment (msf) [inline dune_client/client_async.py:470]: I've seen this if or similar a few times, can it be part of `_build_parameters()` instead ?

File: dune_client/api/base.py [hunk 7]
```diff
@@ -83,6 +83,55 @@ def default_headers(self) -> Dict[str, str]:
...
+        This is shared between the sync and async client.
+        """
+        # Ensure we don't specify parameters that are incompatible:
+        assert (
+            # We are not sampling
+            sample_count is None
+            # We are sampling and don't use filters or pagination
+            or (limit is None and offset is None and filters is None)
+        ), "sampling cannot be combined with filters or pagination"
+
```
Comment (msf) [inline dune_client/api/base.py:112]: I would put here the:
```
        if sample_count is not None:
            limit = None
            offset = None
        else:
            limit = batch_size or MAX_NUM_ROWS_PER_BATCH
            offset = offset or 0
```
or something similar.

**Review by msf — APPROVED** [review]
minor improvements suggested.
1. fix link to API docs (updated)
2. add link to API docs on public function docstrings
3. default values for sample+limit+offset on _build_parameters()

**Pull request merged**